cns/client/client.go

```diff
@@ -0,0 +1,402 @@
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/Azure/azure-container-networking/cns"
+	"github.com/Azure/azure-container-networking/cns/restserver"
+	"github.com/Azure/azure-container-networking/cns/types"
+	"github.com/pkg/errors"
+)
+
+const (
+	contentTypeJSON = "application/json"
+	defaultBaseURL  = "http://localhost:10090"
+	// DefaultTimeout default timeout duration for CNS Client.
+	DefaultTimeout    = 5 * time.Second
+	headerContentType = "Content-Type"
+)
+
+var clientPaths = []string{
+	cns.GetNetworkContainerByOrchestratorContext,
+	cns.CreateHostNCApipaEndpointPath,
+	cns.DeleteHostNCApipaEndpointPath,
+	cns.RequestIPConfig,
+	cns.ReleaseIPConfig,
+	cns.PathDebugIPAddresses,
+	cns.PathDebugPodContext,
+	cns.PathDebugRestData,
+}
+
+// Client specifies a client to connect to Ipam Plugin.
+type Client struct {
+	client http.Client
+	routes map[string]url.URL
+}
+
+// New returns a new CNS client configured with the passed URL and timeout.
+func New(baseURL string, requestTimeout time.Duration) (*Client, error) {
```

Comment:
This may be a larger change, but what we really need to pass into this client is an `*http.Client`. Users of this code should have complete control on all the knobs available on the transport layer. If this is too big to address now, can you add a todo?

Reply:
it's coming, movement in that direction: https://github.com/Azure/azure-container-networking/pull/1001